File: examples/review-diff-agent/src/review_diff_agent.ts
import { fileURLToPath } from "node:url";
import { agent, providerFromProfile, providerProfileFromFile, type ModelProvider } from "@forma-lang/forma";
import { assertReviewDiffOutput, type ReviewDiffInput, type ReviewDiffOutput } from "./review_diff.forma.js";

const exampleInput: ReviewDiffInput = {
  diff: "example",
  max_findings: 1,
};

export function reviewDiffAgent(provider?: ModelProvider) {
  const defaultProvider = provider ?? providerFromProfile(providerProfileFromFile(fileURLToPath(new URL("../forma.provider.json", import.meta.url))));
  return agent({
    file: fileURLToPath(new URL("../review_diff.forma", import.meta.url)),
    task: "review_diff",
    provider: defaultProvider,
  });
}

export async function runReviewDiff(input: ReviewDiffInput = exampleInput, provider?: ModelProvider): Promise<ReviewDiffOutput> {
  const result = await reviewDiffAgent(provider).run({ ...input });
  if (!result.ok) {
    throw new Error(result.error ?? "Forma review_diff failed");
  }
  return assertReviewDiffOutput(result.output);
}

if (import.meta.url === `file://${process.argv[1]}`) {
  const output = await runReviewDiff();
  console.log(JSON.stringify(output, null, 2));
}

File: examples/review-diff-agent/src/review_diff_agent.py
from dataclasses import asdict
from pathlib import Path

from forma import ModelProvider, agent, provider_from_profile, provider_profile_from_file
from review_diff_forma import ReviewDiffInput, ReviewDiffOutput, assert_review_diff_output


PROJECT_ROOT = Path(__file__).resolve().parent.parent

example_input = ReviewDiffInput.from_dict({
        "diff": "example",
        "max_findings": 1,
    })


def review_diff_agent(provider: ModelProvider | None = None):
    default_provider = provider or provider_from_profile(provider_profile_from_file(PROJECT_ROOT / "forma.provider.json"))
    return agent(
        file=PROJECT_ROOT / "review_diff.forma",
        task="review_diff",
        provider=default_provider,
    )


def run_review_diff(input: ReviewDiffInput = example_input, provider: ModelProvider | None = None) -> ReviewDiffOutput:
    result = review_diff_agent(provider=provider).run(asdict(input))
    if not result.ok:
        raise RuntimeError(result.error or "Forma review_diff failed")
    return assert_review_diff_output(result.output)


if __name__ == "__main__":
    print(run_review_diff())
